doc/api/packages.md

```diff
@@ -612,17 +612,23 @@ The following condition definitions are currently endorsed by Node.js:
 * `"browser"` - any environment which implements a standard subset of global
   browser APIs available from JavaScript in web browsers, including the DOM
   APIs.
+* `"deno"` - indicates a variation for the Deno platform.
 * `"development"` - can be used to define a development-only environment
   entry point. _Must always be mutually exclusive with `"production"`._
 * `"production"` - can be used to define a production environment entry
   point. _Must always be mutually exclusive with `"development"`._
+* `"source"` - indicates the original .js source file without minification
+  or bundling optimizations, useful for development and debugging workflows.
+* `"types"` - can be used by typing systems to resolve the typing file for
+  the given export, possible since the interface should be the same for all
```

Comment:
Explicitly: you need to have nested conditional `types` entrypoints if your conditional `exports` entrypoints have differing shapes (eg, because one is a cjs implementation). Meaning a map like:
```json
{
  "exports": {
    "types": {
      "import": "./types/esm/index.d.mts",
      "default": "./types/cjs/index.d.ts"
    },
    "import": "./dist/esm/index.mjs",
    "default": "./dist/cjs/index.js"
  }
}
```



90% of the time it's better to just let your declaration files live alongside your js files than figure this stuff out yourself, though. There are honestly too many axis exports can reasonably vary on (import vs require, dev vs prod, types or not), and it turns out the "the order of matching conditions in the object is the order lookups are attempted in" is confusing in practice, so removing one of those axis is probably preferable when possible.

Reply:
@weswigham point taken, although for large numbers of variations surely this results in a lot of duplicated typing files? The point of such a condition would be to avoid this. I certainly wouldn't recommend it just yet, and I know there are still the dual mode issues, but for ESM only packages with browser / development and Deno variations it seems like a way to avoid three copies of the typing files to me.

Reply:
Alternatively, perhaps the `"source"` condition could be used for typing? As the basic tsc emit with side-by-side declarations, before the second-pass optimization?